<!DOCTYPE html>
<html lang="en"><head>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-html/tabby.min.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.5.47">

  <meta name="author" content="Alex Sanchez, Ferran Reverter and Esteban Vegas">
  <title>Introduction to Statistical Learning – Model validation and Resampling</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="site_libs/revealjs/dist/theme/quarto.css">
  <link rel="stylesheet" href="css4CU.css">
  <link href="site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Model validation and Resampling</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Alex Sanchez, Ferran Reverter and Esteban Vegas 
</div>
        <p class="quarto-title-affiliation">
            Genetics Microbiology and Statistics Department. University of Barcelona
          </p>
    </div>
</div>

</section>
<section>
<section id="model-performance" class="title-slide slide level1 center">
<h1>Model performance</h1>

</section>
<section id="assessing-model-performance" class="slide level2">
<h2>Assessing model performance</h2>
<div class="font80">
<ul>
<li><p>Error estimation and, in general, performance assessment in predictive models is a complex process.</p></li>
<li><p>A key challenge is that <em>the true error of a model on new data is typically unknown</em>, and using the training error as a proxy leads to an optimistic evaluation.</p></li>
<li><p>Resampling methods, such as <em>cross-validation</em> and <em>the bootstrap</em>, allow us to approximate test error and assess model variability using only the available data.</p></li>
<li><p>What is best it can be proven that, well performed, they provide reliable estimates of a model’s performance.</p></li>
<li><p>This section introduces these techniques and discusses their practical implications in model assessment.</p></li>
</ul>
</div>
</section>
<section id="prediction-generalization-error" class="slide level2">
<h2>Prediction (generalization) error</h2>
<ul>
<li><p>We are interested the prediction or generalization error, the error that will appear when predicting a new observation using a model fitted from some dataset.</p></li>
<li><p>Although we don’t know it, it can be estimated using either the training error or the test error estimators.</p></li>
</ul>
</section>
<section id="training-error-vs-test-error" class="slide level2">
<h2>Training Error vs Test error</h2>
<ul>
<li><p>The test error is the average error that results from using a statistical learning method to predict the response on a new observation, one that was not used in training the method.</p></li>
<li><p>The training error is calculated from the difference among the predictions of a model and the observations used to train it.</p></li>
<li><p>Training error rate often is quite different from the test error rate, and in particular the former can dramatically underestimate the latter.</p></li>
</ul>
</section>
<section id="the-three-errors" class="slide level2">
<h2>The three errors</h2>
<div class="font70">
<ul>
<li><p><strong>Generalization Error</strong>. True expected test error (unknown). No bias <span class="math display">\[\mathcal{E}(f) =  \mathbb{E}_{X_0, Y_0} [ L(Y_0, f(X_0)) ]\]</span></p></li>
<li><p><strong>Test Error Estimator</strong>. Estimate of generalization error. Small bias. <span class="math display">\[\hat{\mathcal{E}}_{\text{test}}=\frac{1}{m} \sum_{j=1}^{m} L(Y_j^{\text{test}},  f(X_j^{\text{test}}))\]</span></p></li>
<li><p><strong>Training Error Estimator</strong>. Measures fit to training data (optimistic). High bias <span class="math display">\[\hat{\mathcal{E}}_{\text{train}}: \frac{1}{n} \sum_{i=1}^{n} L(Y_i^{\text{train}}, f(X_i^{\text{train}}))\]</span></p></li>
</ul>
</div>
<!-- | Measure | Formula | Interpretation | Bias | -->
<!-- |---------------------------|---------------|---------------|---------------| -->
<!-- | **Generalization Error** $\mathcal{E}(f)$ | $\mathbb{E}_{X_0, Y_0} [ L(Y_0, f(X_0)) ]$ | True expected test error (unknown) | None | -->
<!-- | **Test Error Estimator** $\hat{\mathcal{E}}_{\text{test}}$ | $\frac{1}{m} \sum_{j=1}^{m} L(Y_j^{\text{test}}, f(X_j^{\text{test}}))$ | Estimate of generalization error (unbiased) | Low | -->
<!-- | **Training Error Estimator** $\hat{\mathcal{E}}_{\text{train}}$ | $\frac{1}{n} \sum_{i=1}^{n} L(Y_i^{\text{train}}, f(X_i^{\text{train}}))$ | Measures fit to training data (optimistic) | High | -->
<!-- ::: -->
</section>
<section id="training--versus-test-set-performance" class="slide level2">
<h2>Training- versus Test-Set Performance</h2>

<img data-src="https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-05.jpg?height=682&amp;width=964&amp;top_left_y=157&amp;top_left_x=151" class="r-stretch"></section>
<section id="prediction-error-estimates" class="slide level2 scrollable">
<h2>Prediction-error estimates</h2>
<ul>
<li><p>Ideal: a large designated test set. Often not available</p></li>
<li><p>Some methods make a mathematical adjustment to the training error rate in order to estimate the test error rate: <span class="math inline">\(Cp\)</span> statistic, <span class="math inline">\(AIC\)</span> and <span class="math inline">\(BIC\)</span>.</p></li>
<li><p>Instead, we consider a class of methods that</p>
<ol type="1">
<li>Estimate test error by holding out a subset of the training observations from the fitting process, and</li>
<li>Apply learning method to held out observations</li>
</ol></li>
</ul>
</section>
<section id="validation-set-approach" class="slide level2">
<h2>Validation-set approach</h2>
<ul>
<li><p>Randomly divide the available samples into two parts: a <em>training set</em> and a <em>validation or hold-out set</em>.</p></li>
<li><p>The model is fit on the training set, and the fitted model is used to predict the responses for the observations <em>in the validation set</em>.</p></li>
<li><p>The resulting validation-set error provides an estimate of the test error. This is assessed using:</p>
<ul>
<li>MSE in the case of a quantitative response and</li>
<li>Misclassification rate in qualitative response.</li>
</ul></li>
</ul>
</section>
<section id="the-validation-process" class="slide level2">
<h2>The Validation process</h2>

<img data-src="https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-08.jpg?height=187&amp;width=830&amp;top_left_y=310&amp;top_left_x=205" class="r-stretch"><p>A random splitting into two halves: left part is training set, right part is validation set</p>
</section>
<section id="example-automobile-data" class="slide level2">
<h2>Example: automobile data</h2>
<ul>
<li><p>Goal: compare linear vs higher-order polynomial terms in a linear regression</p></li>
<li><p>Method: randomly split the 392 observations into two sets,</p>
<ul>
<li>Training set containing 196 of the data points,</li>
<li>Validation set containing the remaining 196 observations.</li>
</ul></li>
</ul>
</section>
<section id="example-automobile-data-plot" class="slide level2">
<h2>Example: automobile data (plot)</h2>

<img data-src="https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-09.jpg?height=379&amp;width=954&amp;top_left_y=466&amp;top_left_x=140" class="r-stretch"><p>Left panel single split; Right panel shows multiple splits</p>
</section>
<section id="drawbacks-of-the-vs-approach" class="slide level2">
<h2>Drawbacks of the (VS) approach</h2>
<div class="font90">
<ul>
<li><p>In the validation approach, <em>only a subset of the observations</em> -those that are included in the training set rather than in the validation set- are used to fit the model.</p></li>
<li><p>The validation estimate of the test error <em>can be highly variable</em>, depending on which observations are included in the training set and which are included in the validation set.</p></li>
<li><p>This suggests that <em>validation set error may tend to over-estimate the test error for the model fit on the entire data set</em>.</p></li>
</ul>
</div>
</section></section>
<section>
<section id="cross-validation" class="title-slide slide level1 center">
<h1>Cross-Validation</h1>

</section>
<section id="k-fold-cross-validation" class="slide level2">
<h2><span class="math inline">\(K\)</span>-fold Cross-validation</h2>
<ul>
<li><p>Widely used approach for estimating test error.</p></li>
<li><p>Estimates give an idea of the test error of the final chosen model</p></li>
<li><p>Estimates can be used to select best model,</p></li>
</ul>
</section>
<section id="k-fold-cv-mechanism" class="slide level2">
<h2><span class="math inline">\(K\)</span>-fold CV mechanism</h2>
<ul>
<li>Randomly divide the data into <span class="math inline">\(K\)</span> equal-sized parts.</li>
<li>Repeat for each part <span class="math inline">\(k=1,2, \ldots K\)</span>,
<ul>
<li>Leave one part, <span class="math inline">\(k\)</span>, apart.</li>
<li>Fit the model to the combined remaining <span class="math inline">\(K-1\)</span> parts,</li>
<li>Then obtain predictions for the left-out <span class="math inline">\(k\)</span>-th part.</li>
</ul></li>
<li>Combine the results to obtain the crossvalidation estimate of tthe error.</li>
</ul>
</section>
<section id="k-fold-cross-validation-in-detail" class="slide level2">
<h2><span class="math inline">\(K\)</span>-fold Cross-validation in detail</h2>

<img data-src="images/clipboard-2845603259.png" class="quarto-figure quarto-figure-center r-stretch"><div class="font60">
<p>A schematic display of 5-fold CV. A set of n observations is randomly split into fve non-overlapping groups. Each of these ffths acts as a validation set (shown in beige), and the remainder as a training set (shown in blue). The test error is estimated by averaging the fve resulting MSE estimates</p>
</div>
</section>
<section id="the-details" class="slide level2">
<h2>The details</h2>
<div class="font80">
<ul>
<li><p>Let the <span class="math inline">\(K\)</span> parts be <span class="math inline">\(C_{1}, C_{2}, \ldots C_{K}\)</span>, where <span class="math inline">\(C_{k}\)</span> denotes the indices of the observations in part <span class="math inline">\(k\)</span>. There are <span class="math inline">\(n_{k}\)</span> observations in part <span class="math inline">\(k\)</span> : if <span class="math inline">\(N\)</span> is a multiple of <span class="math inline">\(K\)</span>, then <span class="math inline">\(n_{k}=n / K\)</span>.</p></li>
<li><p>Compute <span class="math display">\[
\mathrm{CV}_{(K)}=\sum_{k=1}^{K} \frac{n_{k}}{n} \mathrm{MSE}_{k}
\]</span> where <span class="math inline">\(\mathrm{MSE}_{k}=\sum_{i \in C_{k}}\left(y_{i}-\hat{y}_{i}\right)^{2} / n_{k}\)</span>, and <span class="math inline">\(\hat{y}_{i}\)</span> is the fit for observation <span class="math inline">\(i\)</span>, obtained from the data with part <span class="math inline">\(k\)</span> removed.</p></li>
<li><p><span class="math inline">\(K=n\)</span> yields <span class="math inline">\(n\)</span>-fold or <em>leave-one out cross-validation (LOOCV)</em>.</p></li>
</ul>
</div>
<!-- ## A nice special case! -->
<!-- -   With least-squares linear or polynomial regression, an amazing shortcut makes the cost of LOOCV the same as that of a single model -->
<!--     fit! The following formula holds: -->
<!-- $$ -->
<!-- \mathrm{CV}_{(n)}=\frac{1}{n} \sum_{i=1}^{n}\left(\frac{y_{i}-\hat{y}_{i}}{1-h_{i}}\right)^{2} -->
<!-- $$ -->
<!-- where $\hat{y}_{i}$ is the $i$ th fitted value from the original least -->
<!-- squares fit, and $h_{i}$ is the leverage (diagonal of the "hat" matrix; -->
<!-- see book for details.) This is like the ordinary MSE, except the $i$ th -->
<!-- residual is divided by $1-h_{i}$. -->
<!-- -   LOOCV sometimes useful, but typically doesn't shake up the data -->
<!--     enough. The estimates from each fold are highly correlated and hence -->
<!--     their average can have high variance. -->
<!-- -   a better choice is $K=5$ or 10 . -->
</section>
<section id="auto-data-revisited" class="slide level2">
<h2>Auto data revisited</h2>

<!-- ## True and estimated test MSE for the simulated data -->
<!-- ![](https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-19.jpg?height=419&width=348&top_left_y=292&top_left_x=90) -->
<!-- ![](https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-19.jpg?height=415&width=339&top_left_y=294&top_left_x=446) -->
<!-- ![](https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-19.jpg?height=414&width=335&top_left_y=293&top_left_x=803) -->
<img data-src="images/clipboard-3294277508.png" class="r-stretch"></section>
<section id="issues-with-cross-validation" class="slide level2">
<h2>Issues with Cross-validation</h2>
<ul>
<li>Since each training set is only <span class="math inline">\((K-1) / K\)</span> as big as the original training set, the estimates of prediction error will typically be biased upward. Why?</li>
<li>This bias is minimized when <span class="math inline">\(K=n\)</span> (LOOCV), but this estimate has high variance, as noted earlier.</li>
<li><span class="math inline">\(K=5\)</span> or 10 provides a good compromise for this bias-variance tradeoff.</li>
</ul>
</section>
<section id="cv-for-classification-problems" class="slide level2">
<h2>CV for Classification Problems</h2>
<div class="font80">
<ul>
<li>Divide the data into <span class="math inline">\(K\)</span> roughly equal-sized parts <span class="math inline">\(C_{1}, C_{2}, \ldots C_{K}\)</span>.</li>
</ul>
<!-- $C_{k}$ denotes the indices of the observations in part $k$.  -->
<ul>
<li>There are <span class="math inline">\(n_{k}\)</span> observations in part <span class="math inline">\(k\)</span> and <span class="math inline">\(n_{k}\simeq n / K\)</span>.</li>
<li>Compute <span class="math display">\[
\mathrm{CV}_{K}=\sum_{k=1}^{K} \frac{n_{k}}{n} \operatorname{Err}_{k}
\]</span> where <span class="math inline">\(\operatorname{Err}_{k}=\sum_{i \in C_{k}} I\left(y_{i} \neq \hat{y}_{i}\right) / n_{k}\)</span>.</li>
</ul>
</div>
</section>
<section id="standard-error-of-cv-estimate" class="slide level2">
<h2>Standard error of CV estimate</h2>
<ul>
<li>The estimated standard deviation of <span class="math inline">\(\mathrm{CV}_{K}\)</span> is:</li>
</ul>
<p><span class="math display">\[
\widehat{\mathrm{SE}}\left(\mathrm{CV}_{K}\right)=\sqrt{\frac{1}{K} \sum_{k=1}^{K} \frac{\left(\operatorname{Err}_{k}-\overline{\operatorname{Err}_{k}}\right)^{2}}{K-1}}
\]</span></p>
<ul>
<li>This is a useful estimate, but strictly speaking, not quite valid. Why not?</li>
</ul>
</section>
<section id="why-is-this-an-issue" class="slide level2">
<h2>Why is this an issue?</h2>
<div class="font80">
<ul>
<li><p>In (K)-fold CV, the same dataset is used repeatedly for training and testing across different folds.</p></li>
<li><p>This introduces <strong>correlations</strong> between estimated errors in different folds because each fold’s training set overlaps with others.</p></li>
<li><p>The assumption underlying this estimation of the standard error is that <span class="math inline">\(\operatorname{Err}_{k}\)</span> values are <strong>independent</strong>, which does not hold here.</p></li>
<li><p>The dependence between folds leads to <strong>underestimation</strong> of the true variability in <span class="math inline">\(\mathrm{CV}_K\)</span>, meaning that the reported standard error is likely <strong>too small</strong>, giving a misleading sense of precision in the estimate of the test error.</p></li>
</ul>
</div>
</section>
<section id="cv-right-and-wrong" class="slide level2 scrollable">
<h2>CV: right and wrong</h2>
<!-- - CV needs to be performed the right way beacause it is easy to misunderstand how to do it well. -->
<ul>
<li><p>Consider a classifier applied to some 2-class data:</p>
<ol type="1">
<li>Start with 5000 predictors &amp; 50 samples and find the 100 predictors most correlated with the class labels.</li>
<li>We then apply a classifier such as logistic regression, using only these 100 predictors.</li>
</ol></li>
<li><p>In order to estimate the test set performance of this classifier, <em>¿can we apply cross-validation in step 2, forgetting about step 1?</em></p></li>
</ul>
</section>
<section id="cv-the-wrong-and-the-right-way" class="slide level2">
<h2>CV the Wrong and the Right way</h2>
<div class="font90">
<ul>
<li><p>Applying CV only to Step 2 ignores the fact that in Step 1, the procedure has already used the labels of the training data.</p></li>
<li><p>This is a form of training and <strong>must be included in the validation process</strong>.</p>
<ul>
<li>Wrong way: Apply cross-validation in step 2.</li>
<li>Right way: Apply cross-validation to steps 1 and 2.</li>
</ul></li>
<li><p>This error has happened in many high profile papers, mainly due to a misunderstanding of what CV means and does.</p></li>
</ul>
</div>
</section>
<section id="wrong-way" class="slide level2">
<h2>Wrong Way</h2>

<img data-src="https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-31.jpg?height=503&amp;width=1068&amp;top_left_y=229&amp;top_left_x=101" class="r-stretch"></section>
<section id="right-way" class="slide level2">
<h2>Right Way</h2>

<img data-src="https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-32.jpg?height=499&amp;width=1036&amp;top_left_y=234&amp;top_left_x=112" class="r-stretch"></section></section>
<section>
<section id="the-bootstrap" class="title-slide slide level1 center">
<h1>The Bootstrap</h1>

</section>
<section id="introducing-the-bootstrap" class="slide level2">
<h2>Introducing the <em>Bootstrap</em></h2>
<ul>
<li><p>A flexible and powerful statistical tool that can be used to quantify the uncertainty associated with an estimator or a statistical learning method.</p></li>
<li><p>It can provide estimates of the standard error or confidence intervals for that estimator/method.</p></li>
<li><p>Indeed, it can be applied to any (or most) situations where one needs to deal with variability, that the method can approximate using <em>resampling</em>.</p></li>
</ul>
</section>
<section id="where-does-the-name-came-from" class="slide level2">
<h2>Where does the name came from?</h2>
<ul>
<li><p>The term derives from the phrase “to pull oneself up by one’s bootstraps”, thought to be based on the XVIIIth century book ” <em>The Surprising Adventures of Baron Munchausen</em>”.</p></li>
<li><p>It is not the same as the term “bootstrap” used in computer science meaning to “boot” a computer from a set of core instructions.</p></li>
</ul>
</section>
<section id="origins-of-the-bootstrap" class="slide level2">
<h2>Origins of the bootstrap</h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="images/Efron_y_Hastie.png" width="259"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="images/bootLluisCasals.png" width="218"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="a-simple-example" class="slide level2">
<h2>A simple example</h2>
<ul>
<li>Suppose that we wish to invest a fixed sum of money in two financial assets that yield returns of <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>, respectively, where <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> are random quantities.</li>
<li>We will invest a fraction <span class="math inline">\(\alpha\)</span> of our money in <span class="math inline">\(X\)</span>, and will invest the remaining <span class="math inline">\(1-\alpha\)</span> in <span class="math inline">\(Y\)</span>.</li>
<li>We wish to choose <span class="math inline">\(\alpha\)</span> to minimize the total risk, or variance, of our investment. In other words, we want to minimize <span class="math inline">\(\operatorname{Var}(\alpha X+(1-\alpha) Y)\)</span>.</li>
</ul>
</section>
<section id="example-continued" class="slide level2">
<h2>Example continued</h2>
<ul>
<li><p>The value that minimizes the risk is: <span class="math display">\[
\alpha=\frac{\sigma_{Y}^{2}-\sigma_{X Y}}{\sigma_{X}^{2}+\sigma_{Y}^{2}-2 \sigma_{X Y}}
\]</span> where:</p>
<ul>
<li><p><span class="math inline">\(\sigma_{X}^{2}=\operatorname{Var}(X)\)</span></p></li>
<li><p><span class="math inline">\(\sigma_{Y}^{2}=\operatorname{Var}(Y)\)</span> and</p></li>
<li><p><span class="math inline">\(\sigma_{X Y}=\operatorname{Cov}(X, Y)\)</span>.</p></li>
</ul></li>
</ul>
</section>
<section id="example-continued-1" class="slide level2">
<h2>Example continued</h2>
<ul>
<li>The values of <span class="math inline">\(\sigma_{X}^{2}, \sigma_{Y}^{2}\)</span>, and <span class="math inline">\(\sigma_{X Y}\)</span> are unknown.</li>
<li>We can compute estimates for these quantities, <span class="math inline">\(\hat{\sigma}_{X}^{2}, \hat{\sigma}_{Y}^{2}\)</span>, and <span class="math inline">\(\hat{\sigma}_{X Y}\)</span>, using a data set that contains measurements for <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>.</li>
<li>We can then estimate the value of <span class="math inline">\(\alpha\)</span> that minimizes the variance of our investment using <span class="math display">\[
\hat{\alpha}=\frac{\hat{\sigma}_{Y}^{2}-\hat{\sigma}_{X Y}}{\hat{\sigma}_{X}^{2}+\hat{\sigma}_{Y}^{2}-2 \hat{\sigma}_{X Y}}
\]</span></li>
</ul>
</section>
<section id="example-continued-2" class="slide level2">
<h2>Example continued</h2>

<img data-src="images/bootExample1.png" style="width:100.0%" class="r-stretch quarto-figure-center"><p class="caption">Each panel displays 100 simulated returns for investments <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>. <br> From left to right and top to bottom, the resulting estimates for alpha are<br> 0.576, 0.532, 0.657, and 0.651</p></section>
<section id="example-how-variable-is-hat-alpha" class="slide level2 scrollable">
<h2>Example: How variable is <span class="math inline">\(\hat \alpha\)</span>?</h2>
<div class="font90">
<ul>
<li><p>The standard deviation of <span class="math inline">\(\hat{\alpha}\)</span>, can be estimated as:</p>
<ul>
<li><ol type="1">
<li>Repeat 1 … 1000 times</li>
</ol>
<ul>
<li>1.1 Simulate 100 paired observations of <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>.</li>
<li>1.2 Estimate <span class="math inline">\(\alpha: \hat{\alpha}_{1}, \hat{\alpha}_{2}, \ldots, \hat{\alpha}_{1000}\)</span>.</li>
</ul></li>
<li><ol start="2" type="1">
<li>Compute the standard deviation of <span class="math inline">\(\alpha_i\)</span>, <span class="math inline">\(s_{1000}(\hat{\alpha}_i)\)</span> and use it as an estimator of <span class="math inline">\(SE (\hat{\alpha})\)</span></li>
</ol></li>
<li><ol start="3" type="1">
<li>The sample mean of all <span class="math inline">\(\alpha_i\)</span>, <span class="math inline">\(\overline{\hat{\alpha}_i}\)</span> is also a monte carlo estimate of <span class="math inline">\(\alpha\)</span>, although we are not interested in it.</li>
</ol></li>
</ul></li>
</ul>
</div>
</section>
<section id="example-simlation-results" class="slide level2">
<h2>Example: Simlation results</h2>

<img data-src="images/simulatedAlfaValues.png" style="width:90.0%" class="r-stretch"><div class="font50">
<p>Left: A histogram of the estimates of <span class="math inline">\(\alpha\)</span> obtained by generating 1,000 simulated data sets from the true population. Center: A histogram of the estimates of <span class="math inline">\(\alpha\)</span> obtained from 1,000 bootstrap samples from a single data set. Right: The estimates of <span class="math inline">\(\alpha\)</span> displayed in the left and center panels are shown as boxplots.</p>
<p>For these simulations the parameters were set to <span class="math inline">\(\sigma_{X}^{2}=1, \sigma_{Y}^{2}=1.25\)</span>, and <span class="math inline">\(\sigma_{X Y}=0.5\)</span>, and so we know that the true value of <span class="math inline">\(\alpha\)</span> is 0.6 (indicated by the red line).</p>
</div>
</section>
<section id="example-continued-3" class="slide level2">
<h2>Example continued</h2>
<div class="font70">
<ul>
<li>The mean over all 1,000 estimates for <span class="math inline">\(\alpha\)</span> is</li>
</ul>
<p><span class="math display">\[
\bar{\alpha}=\frac{1}{1000} \sum_{r=1}^{1000} \hat{\alpha}_{r}=0.5996,\quad \text { very close to $\alpha=0.6$.}
\]</span></p>
<ul>
<li><p>The standard deviation of the estimates computed on the simulated samples: <span class="math display">\[
\mathrm{SE}_{Sim}(\hat{\alpha})= \sqrt{\frac{1}{1000-1} \sum_{r=1}^{1000}\left(\hat{\alpha}_{r}-\bar{\alpha}\right)^{2}}=0.083
\]</span></p></li>
<li><p>This gives an idea of the accuracy of <span class="math inline">\(\hat{\alpha}\)</span> : <span class="math inline">\(\mathrm{SE}(\hat{\alpha}) \approx \mathrm{SE}_{Sim}(\hat{\alpha})= 0.083\)</span>.</p></li>
<li><p>So roughly speaking, for a random sample from the population, we would expect <span class="math inline">\(\hat{\alpha}\)</span> to differ from <span class="math inline">\(\alpha\)</span> by approximately 0.08 , on average.</p></li>
</ul>
</div>
</section>
<section id="back-to-the-real-world" class="slide level2">
<h2>Back to the real world</h2>
<ul>
<li><p>The procedure outlined above, Monte Carlo Sampling, cannot be applied, because <em>for real data we cannot generate new samples from the original population</em>.</p></li>
<li><p>However, the bootstrap approach allows us to use a computer to mimic the process of obtaining new data sets, so that we can estimate the variability of our estimate without generating additional samples.</p></li>
</ul>
</section>
<section id="sampling-the-sample-resampling" class="slide level2">
<h2>Sampling the sample = <em>resampling</em></h2>
<ul>
<li><p>Rather than repeatedly obtaining independent data sets from the population, we may obtain distinct data sets by <em>repeatedly sampling observations from the original data set with replacement</em>.</p></li>
<li><p>This generates a list of “bootstrap data sets” of the same size as our original dataset.</p></li>
<li><p>As a result some observations may appear more than once in a given bootstrap data set and some not at all.</p></li>
</ul>
</section>
<section id="resampling-illustrated" class="slide level2">
<h2>Resampling illustrated</h2>

<img data-src="images/1.2-Ensemble_Methods-Slides_insertimage_2.png" class="quarto-figure quarto-figure-center r-stretch" style="width:100.0%"></section>
<section id="boot.-estimate-of-std-error" class="slide level2 smaller">
<h2>Boot. estimate of std error</h2>
<ul>
<li><p>The standard error of <span class="math inline">\(\alpha\)</span> can be approximated by the the standard deviation taken on all of these bootstrap estimates using the usual formula: <span class="math display">\[
\mathrm{SE}_{B}(\hat{\alpha})=\sqrt{\frac{1}{B-1} \sum_{r=1}^{B}\left(\hat{\alpha}^{* r}-\overline{\hat{\alpha^{*}}}\right)^{2}}
\]</span></p></li>
<li><p>This quantity, called <em>bootstrap estimate of standard error</em> serves as an estimate of the standard error of <span class="math inline">\(\hat{\alpha}\)</span> estimated from the original data set. <span class="math display">\[
\mathrm{SE}_{B}(\hat{\alpha}) \approx \mathrm{SE}(\hat{\alpha})
\]</span></p></li>
<li><p>For this example <span class="math inline">\(\mathrm{SE}_{B}(\hat{\alpha})=0.087\)</span>.</p></li>
</ul>
</section>
<section id="a-general-picture-for-the-bootstrap" class="slide level2">
<h2>A general picture for the bootstrap</h2>

<img data-src="https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-45.jpg?height=456&amp;width=1072&amp;top_left_y=248&amp;top_left_x=98" class="r-stretch"></section>
<section id="the-bootstrap-in-general" class="slide level2">
<h2>The bootstrap in general</h2>
<ul>
<li>In more complex data situations, figuring out the appropriate way to generate bootstrap samples can require some thought.</li>
<li>For example, if the data is a time series, we can’t simply sample the observations with replacement (why not?).</li>
<li>We can instead create blocks of consecutive observations, and sample those with replacements. Then we paste together sampled blocks to obtain a bootstrap dataset.</li>
</ul>
</section>
<section id="other-uses-of-the-bootstrap" class="slide level2">
<h2>Other uses of the bootstrap</h2>
<ul>
<li>Primarily used to obtain standard errors of an estimate.</li>
<li>Also provides approximate confidence intervals for a population parameter. For example, looking at the histogram in the middle panel of the Figure on slide 29, the <span class="math inline">\(5 \%\)</span> and <span class="math inline">\(95 \%\)</span> quantiles of the 1000 values is (.43, .72 ).</li>
<li>This represents an approximate <span class="math inline">\(90 \%\)</span> confidence interval for the true <span class="math inline">\(\alpha\)</span>. How do we interpret this confidence interval?</li>
</ul>
</section>
<section id="other-uses-of-the-bootstrap-1" class="slide level2">
<h2>Other uses of the bootstrap</h2>
<ul>
<li><p>Primarily used to obtain standard errors of an estimate.</p></li>
<li><p>Also provides approximate confidence intervals for a population parameter. For example, looking at the histogram in the middle panel of the Figure on slide 29, the <span class="math inline">\(5 \%\)</span> and <span class="math inline">\(95 \%\)</span> quantiles of the 1000 values is (.43, .72 ).</p></li>
<li><p>This represents an approximate <span class="math inline">\(90 \%\)</span> confidence interval for the true <span class="math inline">\(\alpha\)</span>. How do we interpret this confidence interval?</p></li>
<li><p>The above interval is called a Bootstrap Percentile confidence interval. It is the simplest method (among many approaches) for obtaining a confidence interval from the bootstrap.</p></li>
</ul>
</section>
<section id="can-the-bootstrap-estimate-prediction-error" class="slide level2">
<h2>Can the bootstrap estimate prediction error?</h2>
<ul>
<li><p>In cross-validation, each of the <span class="math inline">\(K\)</span> validation folds is distinct from the other <span class="math inline">\(K-1\)</span> folds used for training: there is no overlap. This is crucial for its success. Why?</p></li>
<li><p>To estimate prediction error using the bootstrap, we could think about using each bootstrap dataset as our training sample, and the original sample as our validation sample.</p></li>
<li><p>But each bootstrap sample has significant overlap with the original data. About two-thirds of the original data points appear in each bootstrap sample. Can you prove this?</p></li>
<li><p>This will cause the bootstrap to seriously underestimate the true prediction error. Why?</p></li>
<li><p>The other way around- with original sample = training sample, bootstrap dataset <span class="math inline">\(=\)</span> validation sample - is worse!</p></li>
</ul>
</section>
<section id="removing-the-overlap" class="slide level2">
<h2>Removing the overlap</h2>
<ul>
<li><p>Can partly fix this problem by only using predictions for those observations that did not (by chance) occur in the current bootstrap sample.</p></li>
<li><p>But the method gets complicated, and in the end, cross-validation provides a simpler, more attractive approach for estimating prediction error.</p></li>
</ul>
</section>
<section id="pre-validation" class="slide level2">
<h2>Pre-validation</h2>
<ul>
<li><p>In microarray and other genomic studies, an important problem is to compare a predictor of disease outcome derived from a large number of “biomarkers” to standard clinical predictors.</p></li>
<li><p>Comparing them on the same dataset that was used to derive the biomarker predictor can lead to results strongly biased in favor of the biomarker predictor.</p></li>
<li><p>Pre-validation can be used to make a fairer comparison between the two sets of predictors.</p></li>
</ul>
</section>
<section id="motivating-example" class="slide level2 scrollable">
<h2>Motivating example</h2>
<p>An example of this problem arose in the paper of van’t Veer et al. Nature (2002). Their microarray data has 4918 genes measured over 78 cases, taken from a study of breast cancer. There are 44 cases in the good prognosis group and 34 in the poor prognosis group. A “microarray” predictor was constructed as follows:</p>
<ol type="1">
<li>70 genes were selected, having largest absolute correlation with the 78 class labels.</li>
<li>Using these 70 genes, a nearest-centroid classifier <span class="math inline">\(C(x)\)</span> was constructed.</li>
<li>Applying the classifier to the 78 microarrays gave a dichotomous predictor <span class="math inline">\(z_{i}=C\left(x_{i}\right)\)</span> for each case <span class="math inline">\(i\)</span>.</li>
</ol>
</section>
<section id="results" class="slide level2">
<h2>Results</h2>
<p>Comparison of the microarray predictor with some clinical predictors, using logistic regression with outcome prognosis:</p>
<table class="caption-top">
<thead>
<tr class="header">
<th style="text-align: left;">Model</th>
<th style="text-align: right;">Coef</th>
<th style="text-align: right;">Stand. Err.</th>
<th style="text-align: right;">Z score</th>
<th style="text-align: right;">p-value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Re-use</td>
<td style="text-align: right;"></td>
<td style="text-align: right;"></td>
<td style="text-align: right;"></td>
<td style="text-align: right;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">microarray</td>
<td style="text-align: right;">4.096</td>
<td style="text-align: right;">1.092</td>
<td style="text-align: right;">3.753</td>
<td style="text-align: right;">0.000</td>
</tr>
<tr class="odd">
<td style="text-align: left;">angio</td>
<td style="text-align: right;">1.208</td>
<td style="text-align: right;">0.816</td>
<td style="text-align: right;">1.482</td>
<td style="text-align: right;">0.069</td>
</tr>
<tr class="even">
<td style="text-align: left;">er</td>
<td style="text-align: right;">-0.554</td>
<td style="text-align: right;">1.044</td>
<td style="text-align: right;">-0.530</td>
<td style="text-align: right;">0.298</td>
</tr>
<tr class="odd">
<td style="text-align: left;">grade</td>
<td style="text-align: right;">-0.697</td>
<td style="text-align: right;">1.003</td>
<td style="text-align: right;">-0.695</td>
<td style="text-align: right;">0.243</td>
</tr>
<tr class="even">
<td style="text-align: left;">pr</td>
<td style="text-align: right;">1.214</td>
<td style="text-align: right;">1.057</td>
<td style="text-align: right;">1.149</td>
<td style="text-align: right;">0.125</td>
</tr>
<tr class="odd">
<td style="text-align: left;">age</td>
<td style="text-align: right;">-1.593</td>
<td style="text-align: right;">0.911</td>
<td style="text-align: right;">-1.748</td>
<td style="text-align: right;">0.040</td>
</tr>
<tr class="even">
<td style="text-align: left;">size</td>
<td style="text-align: right;">1.483</td>
<td style="text-align: right;">0.732</td>
<td style="text-align: right;">2.026</td>
<td style="text-align: right;">0.021</td>
</tr>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: right;">Pre-validated</td>
<td style="text-align: right;"></td>
<td style="text-align: right;"></td>
<td style="text-align: right;"></td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: right;"></td>
<td style="text-align: right;"></td>
<td style="text-align: right;"></td>
<td style="text-align: right;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">microarray</td>
<td style="text-align: right;">1.549</td>
<td style="text-align: right;">0.675</td>
<td style="text-align: right;">2.296</td>
<td style="text-align: right;">0.011</td>
</tr>
<tr class="even">
<td style="text-align: left;">angio</td>
<td style="text-align: right;">1.589</td>
<td style="text-align: right;">0.682</td>
<td style="text-align: right;">2.329</td>
<td style="text-align: right;">0.010</td>
</tr>
<tr class="odd">
<td style="text-align: left;">er</td>
<td style="text-align: right;">-0.617</td>
<td style="text-align: right;">0.894</td>
<td style="text-align: right;">-0.690</td>
<td style="text-align: right;">0.245</td>
</tr>
<tr class="even">
<td style="text-align: left;">grade</td>
<td style="text-align: right;">0.719</td>
<td style="text-align: right;">0.720</td>
<td style="text-align: right;">0.999</td>
<td style="text-align: right;">0.159</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pr</td>
<td style="text-align: right;">0.537</td>
<td style="text-align: right;">0.863</td>
<td style="text-align: right;">0.622</td>
<td style="text-align: right;">0.267</td>
</tr>
<tr class="even">
<td style="text-align: left;">age</td>
<td style="text-align: right;">-1.471</td>
<td style="text-align: right;">0.701</td>
<td style="text-align: right;">-2.099</td>
<td style="text-align: right;">0.018</td>
</tr>
<tr class="odd">
<td style="text-align: left;">size</td>
<td style="text-align: right;">0.998</td>
<td style="text-align: right;">0.594</td>
<td style="text-align: right;">1.681</td>
<td style="text-align: right;">0.046</td>
</tr>
</tbody>
</table>
</section>
<section id="idea-behind-pre-validation" class="slide level2">
<h2>Idea behind Pre-validation</h2>
<ul>
<li><p>Designed for comparison of adaptively derived predictors to fixed, pre-defined predictors.</p></li>
<li><p>The idea is to form a “pre-validated” version of the adaptive predictor: specifically, a “fairer” version that hasn’t “seen” the response <span class="math inline">\(y\)</span>.</p></li>
</ul>
</section>
<section id="pre-validation-process" class="slide level2">
<h2>Pre-validation process</h2>

<img data-src="https://cdn.mathpix.com/cropped/2025_02_18_d84fddb1dda73076f5eag-64.jpg?height=760&amp;width=812&amp;top_left_y=139&amp;top_left_x=229" class="r-stretch"></section>
<section id="pre-validation-in-detail-for-this-example" class="slide level2 scrollable">
<h2>Pre-validation in detail for this example</h2>
<ol type="1">
<li>Divide the cases up into <span class="math inline">\(K=13\)</span> equal-sized parts of 6 cases each.</li>
<li>Set aside one of parts. Using only the data from the other 12 parts, select the features having absolute correlation at least .3 with the class labels, and form a nearest centroid classification rule.</li>
<li>Use the rule to predict the class labels for the 13th part</li>
<li>Do steps 2 and 3 for each of the 13 parts, yielding a “pre-validated” microarray predictor <span class="math inline">\(\tilde{z}_{i}\)</span> for each of the 78 cases.</li>
<li>Fit a logistic regression model to the pre-validated microarray predictor and the 6 clinical predictors.</li>
</ol>
</section>
<section id="the-bootstrap-versus-permutation-tests" class="slide level2">
<h2>The Bootstrap versus Permutation tests</h2>
<ul>
<li><p>The bootstrap samples from the estimated population, and uses the results to estimate standard errors and confidence intervals.</p></li>
<li><p>Permutation methods sample from an estimated null distribution for the data, and use this to estimate p-values and False Discovery Rates for hypothesis tests.</p></li>
<li><p>The bootstrap can be used to test a null hypothesis in simple situations. Eg if <span class="math inline">\(\theta=0\)</span> is the null hypothesis, we check whether the confidence interval for <span class="math inline">\(\theta\)</span> contains zero.</p></li>
<li><p>Can also adapt the bootstrap to sample from a null distribution (See Efron and Tibshirani book “An Introduction to the Bootstrap” (1993), chapter 16) but there’s no real advantage over permutations.</p></li>
</ul>

<div class="quarto-auto-generated-content">
<div class="footer footer-default">

</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="site_libs/revealjs/plugin/search/search.js"></script>
  <script src="site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true,"width":"half","numbers":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    

    <script>

      // htmlwidgets need to know to resize themselves when slides are shown/hidden.

      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current

      // slide changes (different for each slide format).

      (function () {

        // dispatch for htmlwidgets

        function fireSlideEnter() {

          const event = window.document.createEvent("Event");

          event.initEvent("slideenter", true, true);

          window.document.dispatchEvent(event);

        }

    

        function fireSlideChanged(previousSlide, currentSlide) {

          fireSlideEnter();

    

          // dispatch for shiny

          if (window.jQuery) {

            if (previousSlide) {

              window.jQuery(previousSlide).trigger("hidden");

            }

            if (currentSlide) {

              window.jQuery(currentSlide).trigger("shown");

            }

          }

        }

    

        // hookup for slidy

        if (window.w3c_slidy) {

          window.w3c_slidy.add_observer(function (slide_num) {

            // slide_num starts at position 1

            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);

          });

        }

    

      })();

    </script>

    

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        // For code content inside modals, clipBoardJS needs to be initialized with a container option
        // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>